### DISCUSSION TITLE
GraphCast model integration

### DISCUSSION THREAD

**Discussion by Ker-ple** [Pirate-Weather/pirateweather#183] [Ideas]
[GraphCast](https://github.com/google-deepmind/graphcast) is a weather model released by Google DeepMind which is far more accurate than traditional weather models and has decent accuracy as far as 10 days out. A pre-trained model is [open-sourced](https://github.com/google-deepmind/graphcast) on GitHub and can forecast on a small machine, so it seems possible to integrate with PirateWeather. What do y'all think?

**Comment by alexander0042** [comment]
I saw this earlier, but it's interesting to see the progress on it, and would love to get it rolled in! My big question is how fast it is at running the operational model, since that would probably be the best way to implement it? It looks like the operational one can pull in data and produce a forecast, which is perfect, but if it's heavy to run that would be an issue. 

The other option is to create a forecast for each requested point "live", as it's needed. This would work if it's a really fast model, but would require some sort of automated way to pull in the input data. Either way, it's a super cool project, and definitely on my radar.

**Reply by Ker-ple** [reply]
The operational model is quite slow (minutes) for predicting the first 6-hour step, but quickly speeds up for successive steps.

Because the model makes a prediction for 1440 * 721 = 1038240 points on Earth on each step, I'd recommend:
1. Spin up a Fargate machine to make predictions whenever new data is available.
2. Partition the result into files by geographic location and send the files to S3.
3. Send the files from S3 to CloudFront.
4. Whenever a user makes a request, return the geographically relevant files from the CloudFront cache.

This should minimize time spent processing data and return results to users as quickly as possible.

**Reply by cloneofghosts** [reply]
Just wanted to point out that GFS Graphcast data is available on [AWS](https://registry.opendata.aws/noaa-nws-graphcastgfs-pds/) so you could always integrate that first before hosting the model yourself

**Reply by cloneofghosts** [reply]
@alexander0042 Looks like someone integrated GFS Graphcast into Herbie https://herbie.readthedocs.io/en/latest/user_guide/tutorial/model_notebooks/gfs.html#GFS-GraphCast but it's not in the latest release atm

**Comment by cloneofghosts** [comment]
@alexander0042 Seems like NOAA added more AI models specifically these ones:

- [AIGFS (Artificial Intelligence Global Forecast System)](https://nomads.ncep.noaa.gov/pub/data/nccf/com/aigfs/prod/)
- [AIGEFS (Artificial Intelligence Global Ensemble Forecast System)](https://nomads.ncep.noaa.gov/pub/data/nccf/com/aigefs/prod/)
- [HGEFS (Hybrid-GEFS)](https://nomads.ncep.noaa.gov/pub/data/nccf/com/hgefs/prod/)

Looks like someone created a PR and added them to herbie here: https://github.com/blaylockbk/Herbie/pull/503

**Reply by cloneofghosts** [reply]
This PR https://github.com/Pirate-Weather/pirate-weather-code/pull/507 created to add support for above models.

**Comment by cloneofghosts** [comment]
Ingest has been added for the AIGFS, AIGEFS and AIFS models with it being added to the API response in a future version.
